File: hosts/libretro/cartridge_media.cpp
#include "cartridge_media.h"

#include <type_traits>
#include <variant>

namespace bmsx {

CartridgeCardMedia cartridgeMediaFromPackage(const CartridgePackage& package) {
	CartridgeCardMedia media;
	for (const CartridgeDeviceConfig& device : package.manifest.hardware) {
		std::visit([&media, &package](const auto& concrete) {
			using Config = std::remove_cvref_t<decltype(concrete)>;
			if constexpr (std::is_same_v<Config, CartridgeRomDeviceConfig>) {
				media.rom = package.bytes;
			} else if constexpr (std::is_same_v<Config, CartridgeRamDeviceConfig>) {
				media.ramByteCount = concrete.bytes;
			} else {
				static_assert(std::is_same_v<Config, CartridgeMailboxDeviceConfig>);
				media.mailboxPresent = true;
			}
		}, device);
	}
	return media;
}

} // namespace bmsx

File: hosts/common/cartridge_media.ts
import type { CartridgeCardMedia, CartridgeSocketMediaPair } from '../../machine/ts/machine/devices/cartridge/contracts';
import { parseCartridgePackage, type CartridgePackage } from '../../machine/ts/rompack/image';

/** Decode immutable socket images at host construction, not in IDE features. */
export function cartridgeMediaFromImages(images: readonly [Uint8Array | null, Uint8Array | null]): CartridgeSocketMediaPair {
	const media: CartridgeSocketMediaPair = [null, null];
	for (let slot = 0; slot < images.length; slot += 1) {
		const bytes = images[slot];
		if (bytes !== null) media[slot] = cartridgeMediaFromPackage(parseCartridgePackage(bytes));
	}
	return media;
}

export function cartridgeMediaFromPackage(image: CartridgePackage): CartridgeCardMedia {
	let rom: Uint8Array | null = null;
	let ramByteCount: number | null = null;
	let mailboxPresent = false;
	for (let index = 0; index < image.manifest.hardware.length; index += 1) {
		const device = image.manifest.hardware[index]!;
		switch (device.type) {
		case 'rom':
			rom = image.bytes;
			continue;
		case 'ram':
			ramByteCount = device.bytes;
			continue;
		case 'mailbox':
			mailboxPresent = true;
			continue;
		}
		const unhandledDevice: never = device;
		void unhandledDevice;
	}
	return { rom, ramByteCount, mailboxPresent };
}
